### PR TITLE
keybinding concept

### PR DESCRIPTION
**Pull request by danielsteinkogler**
## Basics

- [x] I added a line to [changelog.md](/doc/changelog.md)
- [x] The PR is rebased with current master.
- [x] Details of what you changed are in commit messages.
- [ ] References to issues, e.g. `close #X`, are in the commit messages and changelog.
- [x] The buildserver is happy.

## Checklist

- [ ] I have installed and I am using [pre-commit hooks](../doc/contrib/README.md#Hooks)
- [ ] I fully described what my PR does in the documentation
- [ ] I fixed all affected documentation
- [ ] I fixed the introduction tour
- [ ] I wrote migrations in a way that they are compatible with already present data
- [ ] I fixed all affected decisions
- [ ] I added automated tests or a [manual test protocol](../doc/tests/manual/protocol.md)
- [ ] I added code comments, logging, and assertions as appropriate
- [ ] I translated all strings visible to the user
- [ ] I mentioned [every code or binary](https://github.com/ElektraInitiative/PermaplanT/blob/master/.reuse/dep5) not directly written or done by me in [reuse syntax](https://reuse.software/)
- [ ] I created left-over issues for things that are still to be done
- [ ] Code is conforming to [our Architecture](/doc/architecture)
- [ ] Code is conforming to [our Guidelines](/doc/guidelines)
- [ ] Code is consistent to [our Design Decisions](/doc/decisions)
- [ ] Exceptions to any guidelines are documented

## Review

- [ ] I've tested the code
- [ ] I've read through the whole code
- [ ] I've read through the whole documentation
- [ ] I've checked conformity to guidelines

### CODE REVIEW

**Comment by danielsteinkogler** [comment]
I added a new concept for our keybinding solution that comes a little bit closer to the solution @tschawax described in https://github.com/ElektraInitiative/PermaplanT/pull/944

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:13]: This is not needed. It should only be developer-friendly: developers should easily get to know where which key binding is assigned.

Having all features to great good UX/UI is however needed (e.g. scoping, maybe @tschawax has more input on that).

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:15]: Before Deciding, give solutions (see template)

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:32]: Please one sentence per line to easier comment on individual statements:

```suggestion
Components that require keybindings can define action handlers for the configured actions.
Here's an example of how this can be implemented:
```

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:52]: What we want from UI perspective should be described in @tschawax document (#918)

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:65]: Put both up to solutions. I agree we can rule out react-hotkeys, as scoping is important from UI perspective.

```suggestion
### React Hotkeys Hook

[react-hotkeys-hook](https://github.com/JohannesKlauss/react-hotkeys-hook/tree/main)

- It provides a hook for binding keys in components and also
- offers scoping abilities.

### React Hotkeys

 [react-hotkeys](https://www.npmjs.com/package/react-hotkeys): provides a component that receives a keymap and actionhandlers similar to our planned concept and listens to events if a child of the component is in focus.

- It seems to be very suitable for encapsulating events, however
- it doesn't really solve the challenge of focus konva layers and
- although the library is widely used the last release was 4 years ago.
```

**Review by markus2330 — CHANGES_REQUESTED** [review]
As the concept is so central to PermaplanT and basically everything but backend is strongly effected, we can discuss it on our regular meeting on Monday.

- https://www.npmjs.com/package/react-hotkeys-hook should be definitely looked into, great point.
- Please also have a look what #1000 is trying to do.

**Comment by danielsteinkogler** [comment]
I restructured the documentation a little bit. I moved everything to solution instead of decision because actually there is no final decision made yet. I look forward to discuss following questions on monday:

- Where to bind keylisteners (especially for konva layers)?
- Should we type available keys and actions (enums?, constants?)?
- Do we want to use https://www.npmjs.com/package/react-hotkeys-hook?

**Comment by danielsteinkogler** [comment]
i updated the concept and moved all refactorings/prove of concept to a seperate PR (https://github.com/ElektraInitiative/PermaplanT/pull/1016)
this PR is currently a draft because i want to add some method documentation tomorrow

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:27]: I wouldn't squeeze that in the same decision. It probably doesn't need a decision but just a way that works nicely. (Concept)

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:29]: Please create a subsubsection for each possible solution.

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:42]:
```suggestion
A JSON file contains all keybindings that are used in the application and serves the following purpose
```

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:44]: Consistency within itemize

```suggestion
- Keybindings can easily be changed by editing the file.
```

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:57]: Can we use some standard names? Toolkits usually provide such names.

```suggestion
    "exitPlantingMode": "Esc"
    "search":{
      "clearSearch": "Esc"
```

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:67]:
```suggestion
### Keyhandling on Focused Elements
```

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:70]:
```suggestion
- method: add scope to JSON, bind a keyhandler and execute action according to configuration
```

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:54]: How can we distinguish layer names, function names etc. here?

File: doc/decisions/frontend_keyhandling.md
Comment (markus2330) [inline doc/decisions/frontend_keyhandling.md:83]: move this concept to guidelines or solution chapter of architecture.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Thx for the improvements. Currrently it is still not a clear decision. For a decision several possibilities should be described, and the best chosen (which packages were not used and why etc.). For some parts of the "decision" it is probably better if you directly write a guideline/concept for how it should work.

File: doc/guidelines/frontend-keyhandling.md
Comment (markus2330) [inline doc/guidelines/frontend-keyhandling.md:76]: `useKeyHandlers` is a function from you? Please introduce it here shortly or at least add comments in the example.

In #1016 only createKeyBindingsAccordingToConfig exists.

Please also make sure the snippets are formatted correctly.

File: doc/decisions/frontend_keybinding.md
Comment (markus2330) [inline doc/decisions/frontend_keybinding.md:1]: Rename file to frontend_keybinding.md

**Review by markus2330 — CHANGES_REQUESTED** [review]
Looks very good to me. :rocket: Only formatting/stylistic changes required. Terms are a bit inconsistent, but I think it should be easy to fix as they only seem to be left-overs.

File: doc/decisions/frontend_keybinding.md
Comment (absurd-turtle) [inline doc/decisions/frontend_keybinding.md:29]:
```suggestion
- however, integration seems not to be as straightforward as with a custom solution and so it doesn't really solve a problem for us.
```

File: doc/decisions/frontend_keybinding.md
Comment (absurd-turtle) [inline doc/decisions/frontend_keybinding.md:38]:
```suggestion
- The custom hook gives us more flexibility and is easier to integrate into our components.
```

File: doc/guidelines/frontend-keyhandling.md
Comment (absurd-turtle) [inline doc/guidelines/frontend-keyhandling.md:32]: I'd suggest swapping the key and the value here if it is not too much effort:
"<<key_binding>>": "<<action_name>>"
The reason is that I think it is easier to look up the corresponding action within the event handler: e.g. config.global['Escape'] to get the corresponding action.
It is also easier to check if there are keybindings assigned multiple times as the json is not valid anymore.

**Review by absurd-turtle — CHANGES_REQUESTED** [review]
Looks pretty good to me :)
I would change the structure of the keybinding config file but other than that everything sounds reasonable.

File: doc/guidelines/frontend-keyhandling.md
Comment (markus2330) [reply doc/guidelines/frontend-keyhandling.md:32]: I also suggested that already but everyone else found it more clear as it is proposed now. A check if key-bindings are used several times should be quite simple in both cases.

> config.global['Escape']

This looks to me to a reason to keep it as proposed. The code shouldn't bother about which keybinding is used, otherwise you couldn't simply make changes in the JSON. Or did I misunderstand?

Comment (absurd-turtle) [reply doc/guidelines/frontend-keyhandling.md:32]: Fair enough. I don't have a strong opinion about it, it is just more intuitive for me as in my experience keybindings are often defined with binding -> action in my experience.
For example vim config, sxhkd config but that doesn't necessarily mean that it is better for us :)
It is not hard to check for duplicates in the config file but my point was that you have to do it.
When the keybinding is the key we have an invalid json as soon as the keybinding is used twice.
On the other hand it might be quicker to look up the bindings from the config with the action as key so we can leave it like that :)

File: doc/guidelines/frontend-keyhandling.md
Comment (danielsteinkogler) [reply doc/guidelines/frontend-keyhandling.md:76]: useKeyHandlers is the custom hook we introduced in a earlier PR that is already merged. i will add a comment.

**Comment by danielsteinkogler** [comment]
jenkins build please

**Comment by danielsteinkogler** [comment]
jenkins build please

**Comment by danielsteinkogler** [comment]
@markus2330 i think we can merge this after approval

File: doc/guidelines/frontend-keyhandling.md
Comment (markus2330) [reply doc/guidelines/frontend-keyhandling.md:5]: There are several change requests still open and also rebase is required.

**Review by markus2330 — CHANGES_REQUESTED** [review]

**Review by absurd-turtle — APPROVED** [review]

File: doc/guidelines/frontend-keyhandling.md
Comment (danielsteinkogler) [reply doc/guidelines/frontend-keyhandling.md:5]: oh sorry for some reason i didn't notice that there a further change requests. i will take care of them now

File: doc/guidelines/frontend-keyhandling.md
Comment (danielsteinkogler) [reply doc/guidelines/frontend-keyhandling.md:7]: i added a checkbox/todo here: https://github.com/ElektraInitiative/PermaplanT/issues/1031.
For implementing this feature I need to check if a layer is active. After implementing this, I will provide specific guidelines on how to do it

File: doc/guidelines/frontend-keyhandling.md
Comment (danielsteinkogler) [reply doc/guidelines/frontend-keyhandling.md]: i added an example to make clear what i meant

**Comment by danielsteinkogler** [comment]
@markus2330 i rebased the pr and resolved the pr comments. it still says that the requested changes have to be addressed. However, i don't see any unresolved comments. Is there a view where i can see only unresolved comments in github?

**Comment by markus2330** [comment]
> @markus2330 i rebased the pr and resolved the pr comments. it still says that the requested changes have to be addressed.
> Or do you have to approve the changes?

Yes, the normal way is that you rerequest reviews when everything is done. :rocket: 

> However, i don't see any unresolved comments. Is there a view where i can see only unresolved comments in github?

I personally use E-Mail notification and go through them one by one.

https://github.com/ElektraInitiative/PermaplanT/pull/994/files might sometimes be helpful,  it shows unresolved comments, but only those who refer to lines that exist. We can discuss such workflows on Monday.

You self-conflicted in the doc/changelog.md. Try a better random generator :wink:

File: doc/guidelines/frontend-keyhandling.md
Comment (markus2330) [inline doc/guidelines/frontend-keyhandling.md:7]:
```suggestion
The file `config/keybindings/keybindings.json` contains all keybindings that are used in the application and serves the following characteristics:
```

Please focus a bit more

File: doc/guidelines/frontend-keyhandling.md
Comment (markus2330) [inline doc/guidelines/frontend-keyhandling.md:10]: These are not really guidelines?

```suggestion
```

File: doc/guidelines/frontend-keyhandling.md
Comment (markus2330) [inline doc/guidelines/frontend-keyhandling.md:12]: (Also missing at other places)

```suggestion
  - `global`: keybindings that are globally active
```

File: doc/guidelines/frontend-keyhandling.md
Comment (markus2330) [inline doc/guidelines/frontend-keyhandling.md:20]: How do these map on Mac OS?
Can you maybe introduce a wildcard modifier key so that it automatically works on Mac OS, too?

File: doc/guidelines/frontend-keyhandling.md
Comment (markus2330) [inline doc/guidelines/frontend-keyhandling.md:8]:
```suggestion

- for all features add commonly-used and vim-like bindings
- we generally try to provide several keybindings for the same functionality, to make the functionality easier accessible
```

File: doc/guidelines/frontend-keyhandling.md
Comment (markus2330) [inline doc/guidelines/frontend-keyhandling.md:42]: First describe what should be done in any case. Ideally add a step-by-step description what someone should do to properly add a keybinding. Then add info about special cases.

**Review by markus2330 — CHANGES_REQUESTED** [review]
Much better, a bit more rework is still needed though.

File: doc/guidelines/frontend-keyhandling.md
Comment (markus2330) [reply doc/guidelines/frontend-keyhandling.md]: Ok, the macOS thing, if it is not trivial/obvious, can also be done with #1031.

**Comment by markus2330** [comment]
I removed "please merge" as I added comments, please readd when you are done (left-over tasks simply to an left-over issue).

File: doc/guidelines/frontend-keyhandling.md
Comment (danielsteinkogler) [reply doc/guidelines/frontend-keyhandling.md:20]: the react key event supports the four mentioned modifier keys (Shift, Alt, Ctrl, Meta). 
The keys should be mapped like this:

- event.ctrlKey: Represents whether the Control key (Ctrl) is pressed.
- event.shiftKey: Represents whether the Shift key is pressed.
- event.altKey: Represents whether the Alt key is pressed (Option key on macOS).
- event.metaKey: Represents whether the Meta key (Command key on macOS) is pressed.

Unfortunately, i cannot test if it is really working like this. I created a left over issue: https://github.com/ElektraInitiative/PermaplanT/issues/1057

File: doc/guidelines/frontend-keyhandling.md
Comment (danielsteinkogler) [reply doc/guidelines/frontend-keyhandling.md:42]: in my opinion the subsections are not describing special cases. there are just two different ways we have to handle keyhandling for either focused elements or keybindings that should be available independent of focus. 
i created a more detailed step-by-step description for both variants.

**Pull request merged**
